File: frontend/lib/permissions.ts
/**
 * FBOS-009: Frontend Route Guards & Permission Helpers
 *
 * Maps roles to permission sets and provides utilities for
 * checking permissions in components and protecting routes.
 */

// ── Role types ──────────────────────────────────────────────────

export type UserRole =
  | "platform_admin"
  | "gym_owner"
  | "manager"
  | "trainer"
  | "dietitian"
  | "customer";

export interface PermissionUser {
  role: string;
  permissions?: string[];
}

// ── Permission matrix ───────────────────────────────────────────

export const ROLE_PERMISSIONS: Record<string, Set<string>> = {
  platform_admin: new Set(["*"]),
  gym_owner: new Set(["*"]),
  manager: new Set([
    "branches.view_branch",
    "customers.view_customer",
    "customers.create_customer",
    "customers.edit_customer",
    "users.view_user",
    "users.create_user",
    "users.edit_user",
    "memberships.view_membership",
    "memberships.create_membership",
    "memberships.edit_membership",
    "payments.view_payment",
    "payments.record_payment",
    "attendance.view_attendance",
    "attendance.log_attendance",
    "dashboard.view_dashboard",
    "reports.view_report",
  ]),
  trainer: new Set([
    "customers.view_customer",
    "memberships.view_membership",
    "attendance.view_attendance",
    "attendance.log_attendance",
    "workouts.view_workout",
    "workouts.create_workout",
    "workouts.edit_workout",
    "dashboard.view_dashboard",
  ]),
  dietitian: new Set([
    "customers.view_customer",
    "diets.view_diet",
    "diets.create_diet",
    "diets.edit_diet",
  ]),
  customer: new Set([
    "memberships.view_membership",
    "payments.view_payment",
    "attendance.view_attendance",
    "attendance.log_attendance",
    "workouts.view_workout",
    "diets.view_diet",
  ]),
};

// ── Permission helpers ──────────────────────────────────────────

export function hasPermission(userRole: string, permission: string): boolean {
  const perms = ROLE_PERMISSIONS[userRole];
  if (!perms) return false;
  if (perms.has("*")) return true;
  return perms.has(permission);
}

export function hasAnyPermission(userRole: string, permissions: string[]): boolean {
  return permissions.some((p) => hasPermission(userRole, p));
}

export function hasAllPermissions(userRole: string, permissions: string[]): boolean {
  return permissions.every((p) => hasPermission(userRole, p));
}

// ── Route guard map ─────────────────────────────────────────────

export const ROUTE_PERMISSIONS: Record<string, string[]> = {
  "/dashboard": ["dashboard.view_dashboard"],
  "/branches": ["branches.view_branch"],
  "/branches/new": ["branches.create_branch"],
  "/customers": ["customers.view_customer"],
  "/customers/new": ["customers.create_customer"],
  "/memberships": ["memberships.view_membership"],
  "/payments": ["payments.view_payment"],
  "/attendance": ["attendance.view_attendance"],
  "/workouts": ["workouts.view_workout"],
  "/diets": ["diets.view_diet"],
  "/users": ["users.view_user"],
  "/reports": ["reports.view_report"],
  "/settings": ["tenants.edit_tenant"],
};

export function canAccessRoute(userRole: string, pathname: string): boolean {
  const required = ROUTE_PERMISSIONS[pathname];
  if (!required) return true; // no permission needed
  return hasAnyPermission(userRole, required);
}

// ── Role display helpers ────────────────────────────────────────

export const ROLE_LABELS: Record<UserRole, string> = {
  platform_admin: "Platform Admin",
  gym_owner: "Gym Owner",
  manager: "Manager",
  trainer: "Trainer",
  dietitian: "Dietitian",
  customer: "Customer",
};

export const ROLE_COLORS: Record<UserRole, string> = {
  platform_admin: "bg-purple-100 text-purple-800",
  gym_owner: "bg-indigo-100 text-indigo-800",
  manager: "bg-blue-100 text-blue-800",
  trainer: "bg-green-100 text-green-800",
  dietitian: "bg-orange-100 text-orange-800",
  customer: "bg-gray-100 text-gray-800",
};

export function getRoleLabel(role: string): string {
  return ROLE_LABELS[role as UserRole] ?? role;
}

export function getRoleColor(role: string): string {
  return ROLE_COLORS[role as UserRole] ?? "bg-gray-100 text-gray-800";
}

File: backend/apps/permissions/permissions.py
"""DRF permission classes for role-based access control."""

from typing import Any

from rest_framework.permissions import BasePermission, IsAuthenticated


class IsPlatformAdmin(BasePermission):
    """Only platform admins (superusers)."""

    def has_permission(self, request: Any, view: Any) -> bool:
        """Check for platform admin role.

        Args:
            request: The incoming DRF request.
            view: The view being accessed.

        Returns:
            ``True`` if the user is authenticated and a superuser.
        """
        user = request.user
        return bool(user and user.is_authenticated and user.is_superuser)


class RolePermission(BasePermission):
    """Permission class that checks a required permission against the user's role.

    Views declare the required permission via ``required_permission`` class attribute.
    Platform admins and gym owners are granted all tenant-scoped permissions.
    """

    ROLE_PERMISSION_MATRIX: dict[str, set[str] | str] = {
        "platform_admin": "*",
        "gym_owner": "*",
        "manager": {
            "branches.view_branch",
            "customers.view_customer",
            "customers.create_customer",
            "customers.edit_customer",
            "users.view_user",
            "users.create_user",
            "users.edit_user",
            "memberships.view_membership",
            "memberships.create_membership",
            "memberships.edit_membership",
            "payments.view_payment",
            "payments.record_payment",
            "attendance.view_attendance",
            "attendance.log_attendance",
            "dashboard.view_dashboard",
            "reports.view_report",
            "feedback.view_feedback",
            "feedback.create_feedback",
            "feedback.edit_feedback",
            "exercises.view_exercise",
            "exercises.edit_exercise",
            "exercises.delete_exercise",
            "marketplace.view_product",
            "marketplace.edit_product",
            "marketplace.delete_product",
            "marketplace.view_cart",
            "marketplace.edit_cart",
            "marketplace.view_order",
            "marketplace.create_order",
            "marketplace.edit_order",
            "marketplace.delete_order",
            "inventory.view_equipment",
            "inventory.create_equipment",
            "inventory.edit_equipment",
            "inventory.delete_equipment",
            "inventory.view_inventory",
            "inventory.create_inventory",
            "inventory.edit_inventory",
            "inventory.delete_inventory",
            "reviews.view_review",
            "reviews.edit_review",
        },
        "trainer": {
            "customers.view_customer",
            "memberships.view_membership",
            "attendance.view_attendance",
            "attendance.log_attendance",
            "workouts.view_workout",
            "workouts.create_workout",
            "workouts.edit_workout",
            "dashboard.view_dashboard",
            "feedback.view_feedback",
            "feedback.edit_feedback",
            "exercises.view_exercise",
            "exercises.edit_exercise",
            "exercises.delete_exercise",
            "reviews.view_review",
            "reviews.edit_review",
        },
        "dietitian": {
            "customers.view_customer",
            "diets.view_diet",
            "diets.create_diet",
            "diets.edit_diet",
            "ai_nutrition.view",
            "ai_nutrition.create",
            "ai_nutrition.edit",
            "marketplace.view_product",
        },
        "customer": {
            "memberships.view_membership",
            "payments.view_payment",
            "attendance.view_attendance",
            "attendance.log_attendance",
            "workouts.view_workout",
            "diets.view_diet",
            "ai_nutrition.view",
            "feedback.view_feedback",
            "feedback.create_feedback",
            "exercises.view_exercise",
            "customers.view_customer",
            "customers.view_healthprofile",
            "customers.view_fitnessgoal",
            "customers.view_bodymeasurement",
            "marketplace.view_product",
            "marketplace.view_cart",
            "marketplace.edit_cart",
            "marketplace.view_order",
            "marketplace.create_order",
            "reviews.view_review",
            "reviews.create_review",
        },
    }

    METHOD_PERMISSIONS: dict[str, str] = {}

    def has_permission(self, request: Any, view: Any) -> bool:
        """Evaluate whether the user's role grants the required permission.

        Args:
            request: The incoming DRF request.
            view: The view being accessed.

        Returns:
            ``True`` if the user is authorized for the required permission.
        """
        if not request.user.is_authenticated:
            return False
        if request.user.is_superuser:
            return True

        # Check method-specific permission first, fall back to class default
        method = getattr(request, "method", None)
        required = None
        if method is not None:
            required = getattr(view, "method_permissions", {}).get(method)
        if required is None:
            required = getattr(view, "required_permission", None)
        if required is None:
            return True

        allowed = self.ROLE_PERMISSION_MATRIX.get(request.user.role, set())
        if allowed == "*":
            return True

        return required in allowed

    def has_object_permission(self, request: Any, view: Any, obj: Any) -> bool:
        """Object-level authorization check.

        Ensures the object belongs to the user's tenant. Customers can only access
        their own records.

        Args:
            request: The incoming DRF request.
            view: The view being accessed.
            obj: The model instance being accessed.

        Returns:
            ``True`` if the object is accessible to the requesting user.
        """
        if not request.user.is_authenticated or request.user.is_superuser:
            return True

        if hasattr(obj, "tenant_id") and obj.tenant_id is not None:
            if obj.tenant_id != request.user.tenant_id:
                return False

        if request.user.role == "customer":
            if hasattr(obj, "user_id") and obj.user_id == request.user.id:
                return True
            if hasattr(obj, "customer_id"):
                from apps.users.models import User

                try:
                    customer_user_id = getattr(
                        obj,
                        "customer_user_id",
                        User.objects.get(
                            customer_profile__id=obj.customer_id,
                        ).id,
                    )
                except User.DoesNotExist:
                    return False
                return customer_user_id == request.user.id
            return False

        return True


class RolePermissionMixin:
    """Convenience mixin for class-based views that sets permission classes.

    Subclasses must set ``required_permission``.
    """

    required_permission: str | None = None
    authentication_classes = ["apps.users.authentication.TenantTokenAuthentication"]
    permission_classes = [
        IsAuthenticated,
        "apps.tenants.permissions.IsTenantMember",
        RolePermission,
    ]
